src/execution_plan/ops/op_expand_into.c

```diff
@@ -6,68 +6,270 @@
+     * otherwise try to get a new pair of source and destination nodes. */
+    if(op->ae->edge) {
+        if(_setEdge(op)) return Record_Clone(op->r);
+    }
+
+    Node *srcNode;
+    Node *destNode;
+    NodeID srcId = INVALID_ENTITY_ID;
+    NodeID destId = INVALID_ENTITY_ID;
+
+    /* Find a record where both record's source and destination
+     * nodes are connected. */
+    while(op->recordsLen) {
+        op->recordsLen--;
+        op->r = op->records[op->recordsLen];
+
+        srcNode = Record_GetNode(op->r, op->srcNodeRecIdx);
+        destNode = Record_GetNode(op->r, op->destNodeRecIdx);
+        srcId = ENTITY_GET_ID(srcNode);
+        destId = ENTITY_GET_ID(destNode);
+
+        bool x;
+        GrB_Info res = GrB_Matrix_extractElement_BOOL(&x, op->M, destId, srcId);
+        // Src is not connected to dest.
+        if(res != GrB_SUCCESS) continue;
+
+        // If we're here src is connected to dest.
+        if(op->ae->edge) {
+            for(int i = 0; i < op->edgeRelationCount; i++) {
+                Graph_GetEdgesConnectingNodes(op->graph,
+                                            srcId,
+                                            destId,
+                                            op->edgeRelationTypes[i],
+                                            &op->edges);
+            }
+            _setEdge(op);
+            return Record_Clone(op->r);
+        }
+
+        // Mark as NULL to avoid double free.
+        op->records[op->recordsLen] = NULL;
+        return op->r;
+    }
+
+    // Didn't manage to emit record.
+    return NULL;
+}
+
+/* ExpandIntoConsume next operation 
+ * returns OP_DEPLETED when no additional updates are available */
+Record ExpandIntoConsume(OpBase *opBase) {
+    Node *n;
+    Record r;
     OpExpandInto *op = (OpExpandInto*)opBase;
     OpBase *child = op->op.children[0];
-    Record r = NULL;
-
-    // Clear edges array, see if a is connected to b via e.
-    array_clear(op->edges);
-    while(array_len(op->edges) == 0) {
-        // Try to get data, return NULL if depleted.
-        r = child->consume(child);
-        if(r == NULL) return NULL;
-
-        // Get node a and b from record.
-        Node *a = Record_GetNode(r, op->srcRecIdx);
-        Node *b = Record_GetNode(r, op->destRecIdx);
-        assert(a && b);
-
-        Graph_GetEdgesConnectingNodes(
-            op->gc->g,
-            ENTITY_GET_ID(a),
-            ENTITY_GET_ID(b),
-            Edge_GetRelationID(op->e),
-            op->edges
-        );
+
+    // As long as we don't have a record to emit.
+    while((r = _handoff(op)) == NULL) {
+        /* If we're here, we didn't managed to emit a record,
+         * clean up and try to get new data points. */
+        for(int i = 0; i < op->recordsCap; i++) {
+            if(op->records[i]) {
+                Record_Free(op->records[i]);
+                op->records[i] = NULL;
+            } else {
+                break;
```

Comment:
So all subsequent records are guaranteed to be NULL once we encounter the first NULL?

Reply:
As records come in from child operations, we store record I at op->records[i]
until we either unable to get additional records, or we've ran out of space in op->records.
when emitting a records, we scan op->records backwards, setting each emitted record to NULL. with the exception where we required to set the edge connecting `a` to `b` in which case we clone the record and retain the record at op->records[i].

Keep in mind situations where LIMIT is introduced.
And so I believe it is OK to break once we encounter NULL.